sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIOProducer.java

```diff
@@ -0,0 +1,222 @@
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.util.BackOff;
+import org.apache.beam.sdk.util.BackOffUtils;
+import org.apache.beam.sdk.util.FluentBackoff;
+import org.apache.beam.sdk.util.Sleeper;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings({
+  "nullness" // TODO(https://github.com/apache/beam/issues/20497)
+})
+public class JmsIOProducer<T> extends PTransform<PCollection<T>, WriteJmsResult<T>> {
+
+  public static final String CONNECTION_ERRORS_METRIC_NAME = "connectionErrors";
+  public static final String PUBLICATION_RETRIES_METRIC_NAME = "publicationRetries";
+  public static final String JMS_IO_PRODUCER_METRIC_NAME = JmsIOProducer.class.getCanonicalName();
+
+  private static final Logger LOG = LoggerFactory.getLogger(JmsIOProducer.class);
+  private static final String PUBLISH_TO_JMS_STEP_NAME = "Publish to JMS";
+
+  private final JmsIO.Write<T> spec;
+  private final TupleTag<T> messagesTag;
+  private final TupleTag<T> failedMessagesTag;
+
+  JmsIOProducer(JmsIO.Write<T> spec) {
+    this.spec = spec;
+    this.messagesTag = new TupleTag<>();
+    this.failedMessagesTag = new TupleTag<>();
+  }
+
+  @Override
+  public WriteJmsResult<T> expand(PCollection<T> input) {
+    PCollectionTuple failedPublishedMessages =
+        input.apply(
+            PUBLISH_TO_JMS_STEP_NAME,
+            ParDo.of(new JmsIOProducerFn())
+                .withOutputTags(messagesTag, TupleTagList.of(failedMessagesTag)));
+    PCollection<T> failedMessages =
+        failedPublishedMessages.get(failedMessagesTag).setCoder(input.getCoder());
+    failedPublishedMessages.get(messagesTag).setCoder(input.getCoder());
+    return WriteJmsResult.in(input.getPipeline(), failedMessagesTag, failedMessages);
+  }
+
+  private class JmsIOProducerFn extends DoFn<T, T> {
+
+    private transient @Initialized Session session;
+    private transient @Initialized Connection connection;
+    private transient @Initialized Destination destination;
+    private transient @Initialized MessageProducer producer;
+    private transient @Initialized FluentBackoff retryPublicationBackoff;
+
+    private final Counter connectionErrors =
+        Metrics.counter(JMS_IO_PRODUCER_METRIC_NAME, CONNECTION_ERRORS_METRIC_NAME);
+    private final Counter publicationRetries =
+        Metrics.counter(JMS_IO_PRODUCER_METRIC_NAME, PUBLICATION_RETRIES_METRIC_NAME);
+
+    @Setup
+    public void setup() {
+      retryPublicationBackoff =
+          FluentBackoff.DEFAULT
+              .withMaxRetries(0)
```

Comment:
@Amraneze thanks for explanation. I get where the issue is. What I mean previously ""by default" was that, the sink should retry by default. Here I see if retrypolicy is not setup, there is still no retry (current behavior)

I am suggesting follow the same pattern used elsewhere (e.g. https://github.com/apache/beam/blob/aba175259944caa75a7673b9e0ad430d8e5c68fe/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L2459-L2465)

Specifically, name the retry policy as RetryConfiguration that is an internal class of JmsIO, and set a default retry configuration if user does not provide it. I see there is a substantial refactoring of the code in this PR which may not be necessary (producer can still be in JmsIO class) finally thanks that the unit test has been implemented.

Reply:
Ok, I will work on it.

Reply:
@Abacn Can you recheck if everything is good ?